src/clr-core/common/utils/aria-live.service.ts

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { AriaLivePoliteness } from '../enums/aria-live-politeness';
+import { applyMixins } from '../mixins/apply-mixins';
+import { UniqueId } from '../mixins/unique-id';
+
+/**
+ * Time in milliseconds before inserting the content into the container
+ */
+export const ARIA_LIVE_TICK: number = 100;
+export class CwcAriaLiveService {
+  protected previousTimeout: ReturnType<typeof setTimeout>;
+  protected ariaLiveElement: HTMLElement | null;
+  document: Document;
+
+  readonly id = `clr-aria-live-element-${this._uniqueId}`;
+
+  constructor(document: Document) {
+    this.document = document;
+  }
+
+  /**
+   * Append text content inside the AriaLive Container. This method will check if the
+   * DOM Element is existing if not it will create one for us and the will apply the text.
+   *
+   * ```typescript
+   * this.ariaLiveService.announce(this.el.nativeElement);
```

Comment:
The service is design to handle its own cases, and create new DOM Element for every component. So one or more component could broadcast there messages independently and let the SR decide the order. Also different component could have different politeness type. That was the initial idea of going with multiple instances. 

Going static is an option as well, there are not a lot going on inside the constructors. And the idea for having it this way for the angular side was to get the document injected and using the ngOnDestroy auto call when component is destroyed so we only care for creating the message and let the “nature selection” do its thing. 

For using it as static methods there have to be made some changes on how new ids are generated for the creation of aria live containers and how we keep the reference for the required element. Also to be sure that there will be no left behind DOM Elements, this could quickly create a lot of unused elements if there is a for example progress or any type of component that create message quickly.

Reply:
From looking at the code, it appears we are only serendipitously cleaning up after the service anyway. As every use of `announce()` creates a new DOM element with the same id as the previous one tied to the service.

Having multiple elements on the same page with the same id isn't supposed to happen. At the very least we need to make the cleanup more resilient. Looking over the code, I'm not convinced that it isn't leaving stragglers behind. I'd at least want to see a test that (a) uses `announce` on the service multiple times and (b) verifies that we don't have any stragglers in the DOM after cleanup. 

Based on what I'm seeing, I don't have high confidence that test would pass.

Reply:
The method has to create new DOM element only if there is no existing element already. And the ID has to be different for every instance. So in use there will be only one element per component and it will reuse it for the next message. There must be no duplicate IDs. 

At least for the angular side, if we destroy the component we gonna remove the DOM element. This is valid only for the component that is initialize the service. If for some reason, the component fail and throw error or something prevent the code for finishing the ngOnDestroy this may not remove and create orphans.